lib/diskoverflow/common.go

```diff
@@ -0,0 +1,154 @@
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+const concurrencyMsg = "iteration in progress - don't modify or start a new iteration concurrently"
+
+const OrigDefaultOverflowBytes = 16 << protocol.MiB
+
+var (
+	defaultOverflow    = OrigDefaultOverflowBytes
+	defaultOverflowMut = sync.Mutex{}
+)
+
+// Sets the default limit of when data is written to disks. Can be overruled
+// on individual instances via their SetOverflowBytes method.
+// A value of bytes =< 0 means no disk spilling will ever happen.
+func SetDefaultOverflowBytes(bytes int) {
+	defaultOverflowMut.Lock()
+	defaultOverflow = bytes
+	defaultOverflowMut.Unlock()
+}
+
+func DefaultOverflowBytes() int {
+	defaultOverflowMut.Lock()
+	defer defaultOverflowMut.Unlock()
+	return defaultOverflow
+}
+
+// Value must be implemented by every type that is to be stored in a disk spilling container.
+// Un-/Marshal returning an error will result in a panic.
+// This uses methods from proto interfaces such that existing types can be reused.
+type Value interface {
+	Marshal() ([]byte, error)
+	Unmarshal([]byte) error
+	Reset() // To make an already populated Value ready for Unmarshal
+	ProtoSize() int
+}
+
+// copyValue copies the content from src to dst. Src and dst must be pointers
+// to the same underlying types, otherwise this will panic.
+func copyValue(dst, src Value) {
+	dstv := reflect.ValueOf(dst).Elem()
+	srcv := reflect.ValueOf(src).Elem()
+	dstv.Set(srcv)
+}
+
+type common interface {
+	Bytes() int
+	Items() int
+	Close()
+}
+
+type base struct {
+	location      string
+	overflowBytes int
+	spilling      bool
+	iterating     bool
+}
+
+func newBase(location string) base {
+	defaultOverflowMut.Lock()
+	defer defaultOverflowMut.Unlock()
+	return base{
+		location:      location,
+		overflowBytes: defaultOverflow,
+	}
+}
+
+func (o *base) SetOverflowBytes(bytes int) {
+	o.overflowBytes = bytes
+}
+
+func (o *base) startSpilling(size int) bool {
+	return !o.spilling && o.overflowBytes > 0 && size > o.overflowBytes
+}
+
+type Iterator interface {
+	Release()
+	Next() bool
+	Value(Value)
+}
+
+type iteratorParent interface {
+	released()
+}
+
+type memIterator struct {
+	values  []Value
+	pos     int
+	len     int
+	reverse bool
+	parent  iteratorParent
+}
+
+func newMemIterator(values []Value, p iteratorParent, reverse bool, len int) *memIterator {
+	it := &memIterator{
+		values:  values,
+		len:     len,
+		reverse: reverse,
+		parent:  p,
+	}
+	if reverse {
+		it.pos = len
+	} else {
+		it.pos = -1
+	}
+	return it
+}
+
+func (si *memIterator) Next() bool {
+	if si.reverse {
+		if si.pos == 0 {
+			return false
+		}
+		si.pos--
+		return true
+	}
+	if si.pos == si.len-1 {
+		return false
+	}
+	si.pos++
+	return true
+}
+
+func (si *memIterator) Value(v Value) {
+	if si.pos != si.len && si.pos != -1 {
+		copyValue(v, si.values[si.pos])
+	}
+}
+
+func (si *memIterator) Release() {
+	si.parent.released()
+}
+
+func errPanic(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
```

Comment:
make mustMarshal(), mustUnmarshal() perhaps.

Reply:
Also, as a library user, it wouldn't be cool if some data structure library crashed my app because of /tmp filling up or something like that.

Reply:
I agree and since using the marshalling interface there are already lots of errors, which should never happen, so why not introduce a bunch more errors to the API.

As to the anti-pattern: `errPanic` is just a more generic version of `mustMarshal` - what makes the former an anti-pattern and the latter not?

Reply:
I guess that is also an antipattern especially in a context of a library, but at least it's not a generic if not nil panic thing which actually does something.